components/ws-daemon/pkg/controller/workspace_operations.go

```diff
@@ -0,0 +1,492 @@
+	}
+
+	// Update the git status prior to deleting the workspace
+	repo, err = sess.UpdateGitStatus(ctx, false)
+	if err != nil {
+		// do not fail workspace because we were unable to get git status
+		// which can happen for various reasons, including user corrupting his .git folder somehow
+		// instead we log the error and continue cleaning up workspace
+		// todo(pavel): it would be great if we can somehow bubble this up to user without failing workspace
+		glog.WithError(err).Warn("cannot get git status")
+	}
+
+	if err = sess.Dispose(ctx); err != nil {
+		glog.WithError(err).Error("cannot dispose session")
+	}
+
+	// remove workspace daemon directory in the node
+	if err := os.RemoveAll(sess.ServiceLocDaemon); err != nil {
+		glog.WithError(err).Error("cannot delete workspace daemon directory")
+	}
+
+	return false, repo, nil
+}
+
+func (wso *WorkspaceOperations) uploadWorkspaceLogs(ctx context.Context, opts DisposeOptions) (err error) {
+	// currently we're only uploading prebuild log files
+	logFiles, err := logs.ListPrebuildLogFiles(ctx, opts.WorkspaceLocation)
+	if err != nil {
+		return err
+	}
+
+	rs, err := storage.NewDirectAccess(&wso.config.Storage)
+	if err != nil {
+		return xerrors.Errorf("cannot use configured storage: %w", err)
+	}
+
+	err = rs.Init(ctx, opts.Meta.Owner, opts.Meta.WorkspaceId, opts.Meta.InstanceId)
+	if err != nil {
+		return xerrors.Errorf("cannot use configured storage: %w", err)
+	}
+
+	err = rs.EnsureExists(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, absLogPath := range logFiles {
+		taskID, parseErr := logs.ParseTaskIDFromPrebuildLogFilePath(absLogPath)
+		if parseErr != nil {
+			glog.WithError(parseErr).Warn("cannot parse headless workspace log file name")
+			continue
+		}
+
+		err = retryIfErr(ctx, 5, glog.WithField("op", "upload log"), func(ctx context.Context) (err error) {
+			_, _, err = rs.UploadInstance(ctx, absLogPath, logs.UploadedHeadlessLogPath(taskID))
+			if err != nil {
+				return
+			}
+
+			return
+		})
+		if err != nil {
+			return xerrors.Errorf("cannot upload workspace logs: %w", err)
+		}
+	}
+	return err
+}
+
+func (wso *WorkspaceOperations) uploadWorkspaceContent(ctx context.Context, sess *session.Workspace) error {
+	var backupName = storage.DefaultBackup
+	// Avoid too many simultaneous backups in order to avoid excessive memory utilization.
+	waitStart := time.Now()
+	select {
+	case wso.backupWorkspaceLimiter <- struct{}{}:
+	case <-time.After(15 * time.Minute):
+		// we timed out on the rate limit - let's upload anyways, because we don't want to actually block
+		// an upload. If we reach this point, chances are other things are broken. No upload should ever
+		// take this long.
+		wso.metrics.BackupWaitingTimeoutCounter.Inc()
+	}
+
+	waitTime := time.Since(waitStart)
+	wso.metrics.BackupWaitingTimeHist.Observe(waitTime.Seconds())
+
+	defer func() {
+		// timeout -> we did not add to the limiter
+		if waitTime.Minutes() >= 15 {
+			return
+		}
```

Comment:
still slightly prone to a race condition, what about setting a boolean if we reach the `case wso.backupWorkspaceLimiter <- struct{}{}:` case, and return based on that? Or otherwise, lets make `15` a const, so we don't forget to update the other if we change either